IdentityCore/src/network/MSIDHttpRequest.m

```diff
@@ -0,0 +1,117 @@
+// to use, copy, modify, merge, publish, distribute, sublicense, and / or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions :
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+#import "MSIDHttpRequest.h"
+#import "MSIDJsonResponseSerializer.h"
+#import "MSIDUrlRequestSerializer.h"
+#import "MSIDHttpRequestTelemetryProtocol.h"
+#import "MSIDHttpRequestErrorHandlerProtocol.h"
+#import "MSIDHttpRequestConfiguratorProtocol.h"
+#import "MSIDHttpRequestTelemetry.h"
+
+@interface MSIDHttpRequest () <NSURLSessionDelegate>
+
+@property (nonatomic) NSURLSessionConfiguration *sessionConfiguration;
+@property (nonatomic) NSURLSession *session;
+
+@end
+
+@implementation MSIDHttpRequest
+
+- (instancetype)init
+{
+    self = [super init];
+    
+    if (!self) return nil;
+    
+    _sessionConfiguration = [NSURLSessionConfiguration defaultSessionConfiguration];
+    _session = [NSURLSession sessionWithConfiguration:_sessionConfiguration delegate:self delegateQueue:nil];
+    _responseSerializer = [MSIDJsonResponseSerializer new];
+    _requestSerializer = [MSIDUrlRequestSerializer new];
+    _telemetry = [MSIDHttpRequestTelemetry new];
+    
+    return self;
+}
+
+- (void)sendWithBlock:(MSIDHttpRequestDidCompleteBlock)completionBlock;
+{
+    NSParameterAssert(self.urlRequest);
+    
+    self.urlRequest = [self.requestSerializer serializeWithRequest:self.urlRequest parameters:self.parameters];
+    
+    if (self.requestConfigurator) [self.requestConfigurator configure:self];
+    
+    [self.telemetry sendRequestEventWithId:self.context.telemetryRequestId];
+    
+    [[self.session dataTaskWithRequest:self.urlRequest completionHandler:^(NSData *data, NSURLResponse *response, NSError *error)
+      {
+          MSID_LOG_VERBOSE(self.context, @"Received network response: %@, error %@", _PII_NULLIFY(response), _PII_NULLIFY(error));
+          MSID_LOG_VERBOSE_PII(self.context, @"Received network response: %@, error %@", response, error);
+          
+          NSAssert([response isKindOfClass:NSHTTPURLResponse.class], NULL);
+          
+          __auto_type httpResponse = (NSHTTPURLResponse *)response;
+          
+          [self.telemetry responseReceivedEventWithId:self.context.telemetryRequestId
+                                        correlationId:self.context.correlationId
```

Comment:
nit: can we just pass the context which has both telemetryRequestId and correlationId already?

Reply:
done